text/0032-react-primary-framework.md

```diff
@@ -0,0 +1,105 @@
+# Making React the Primary Framework
+
+- Start Date: 2021/11/26
+- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/32
+
+## Decision, including impact on distributions
+
+The app shell should already be developed using React, making React a first-class citizen. As such unnecessary framework-independent packages such as `single-spa` can be removed. Instead, `react-router` will be the central router.
+
+Existing distributions / frontend modules should not be impacted - a compatibility layer will be introduced.
+
+## Reason for decision
+
+Right now every frontend module registers a couple of components (mostly "extensions" and "pages"). Each component will be rendered using a completely new rendering tree. This comes with some significant overhead as compared to reusing a single rendering tree.
+
+Since most of our components (>= 95%) are React components and since in React its very easy to have a single rendering tree dispatching to multiple DOM nodes (using the `createPortal` API of React) we should consider going to a single rendering tree of React. The other advantage is that the framework agnostic approach has problems with otherwise simple things such as prop updates. While this is fully integrated and highly optimized in React, the framework agnostic approach right now makes the update inefficient.
+
+Besides the performance part we would also see significant reduction in complexity. Right now developers need to think in multiple dimensions. For instance, the routing is happening on an *outside* level (using `single-spa`s router) and potentially even on an *inside* level (using, e.g., the `react-router`). While most of the time this works as expected, sometimes weird behavior is observed.
+
+## Alternatives
+
+There are some ways we can go alternatively, including:
+
+1. As-is, but with enhancements to the `single-spa-react` connector (would have several downsides and would not provide all the optimizations we look for)
```

Comment:
For instance still having a router of router, which degrades developer and user experience.

Reply:
```suggestion
1. Make enhancements to the `single-spa-react` connector. This would eliminate the unnecessary React render trees and allow React to handle props directly. Routing, however, would remain as it is currently.
```